pkg/webhook/server/certificate/manager_test.go

```diff
@@ -0,0 +1,181 @@
+		{"long duration", now.Add(-6 * 30 * 24 * time.Hour), now.Add(6 * 30 * 24 * time.Hour), true},
+		{"short duration", now.Add(-30 * time.Second), now.Add(30 * time.Second), true},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			m := Manager{
+				caKeyPair: &triple.KeyPair{
+					Cert: &x509.Certificate{
+						NotBefore: tc.notBefore,
+						NotAfter:  tc.notAfter,
+					},
+				},
+				now: func() time.Time { return now },
+				log: log,
+			}
+			jitteryDuration = func(float64) time.Duration { return time.Duration(float64(tc.notAfter.Sub(tc.notBefore)) * 0.7) }
+			lowerBound := tc.notBefore.Add(time.Duration(float64(tc.notAfter.Sub(tc.notBefore)) * 0.7))
+
+			deadline := m.nextRotationDeadline()
+
+			if !deadline.Equal(lowerBound) {
+				t.Errorf("For notBefore %v, notAfter %v, the rotationDeadline %v should be %v.",
+					tc.notBefore,
+					tc.notAfter,
+					deadline,
+					lowerBound)
+			}
+		})
+	}
+}
+
+func TestWaitForDeadlineAndRotate(t *testing.T) {
+
+	certDir, err := ioutil.TempDir("/tmp/", "manager-test-certs")
+	if err != nil {
+		t.Errorf("failed creating temporal dir for certs %v", err)
+	}
+	defer os.RemoveAll(certDir)
+
+	mutatingWebhookConfiguration := &admissionregistrationv1beta1.MutatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "fooWebhook",
+		},
+		Webhooks: []admissionregistrationv1beta1.MutatingWebhook{
+			admissionregistrationv1beta1.MutatingWebhook{
+				Name: "fooWebhook",
+				ClientConfig: admissionregistrationv1beta1.WebhookClientConfig{
+					Service: &admissionregistrationv1beta1.ServiceReference{
+						Name:      "fooWebhook",
+						Namespace: "fooWebhook",
+					},
+				},
+			},
+		},
+	}
+
+	objs := []runtime.Object{mutatingWebhookConfiguration}
+
+	client := fake.NewFakeClient(objs...)
+	certsDuration := time.Minute
+	manager := NewManager(client, "fooWebhook", MutatingWebhook, certsDuration)
+	manager.waitForDeadlineAndRotate()
+	//TODO Implement ErrorsHandler to take the errors that we have at
+	//     background
+
+	err = client.Get(context.TODO(), types.NamespacedName{Name: "fooWebhook"}, mutatingWebhookConfiguration)
+	if err != nil {
+		t.Fatalf("get mutatingwebhookconfiguration: (%v)", err)
+	}
+
+	clientConfig := mutatingWebhookConfiguration.Webhooks[0].ClientConfig
+	if len(clientConfig.CABundle) == 0 {
+		t.Error("CA bundle not updated")
+	}
+
+	secret := corev1.Secret{}
+	err = client.Get(context.TODO(), types.NamespacedName{Name: "fooWebhook", Namespace: "fooWebhook"}, &secret)
+	if err != nil {
+		t.Fatalf("get secret: (%v)", err)
+	}
+
+	if secret.Type != corev1.SecretTypeTLS {
+		t.Fatalf("Non TLS secret type %s", secret.Type)
+	}
+
+	if len(secret.Data) == 0 {
+		t.Fatal("No tls key/cert at secret")
+	}
+
+	nextRotation := time.Now().Sub(manager.nextRotationDeadline())
+	start := time.Now()
+	err = wait.PollImmediate(nextRotation, nextRotation, func() (bool, error) {
+		manager.waitForDeadlineAndRotate()
```

Comment:
I don't understand. does this function initiate the rotation? It looks like it from process of elimination, but want to be sure..

Reply:
The deadline is calculated based on two circustances:
- existence
- expiricy

First run is jus the first one.